File: android/app/src/main/kotlin/com/app/deenly/deenly/assetdownload/AssetDownloadError.kt
package com.rnr.deenfocus.assetdownload

/**
 * Generic exceptions for the reusable native asset-download framework (ADR-008).
 * Intentionally independent of any specific consumer's error codes (e.g. Tajweed's
 * `TajweedErrorCode`) so this framework has zero knowledge of AI inference, ONNX,
 * or Tajweed — consumers translate these into their own error contracts at the
 * adapter boundary (see `TajweedAssetSync`).
 */
sealed class AssetDownloadException(message: String) : Exception(message) {
    class Network(message: String) : AssetDownloadException(message)
    class BadStatusCode(val code: Int) : AssetDownloadException("HTTP $code")
    class DecodeFailed(message: String) : AssetDownloadException(message)
    object Timeout : AssetDownloadException("Timed out")
    object Cancelled : AssetDownloadException("Cancelled")
    class InsufficientStorage(val requiredBytes: Long, val availableBytes: Long) :
        AssetDownloadException("Need $requiredBytes bytes, only $availableBytes available")
    class DiskError(message: String) : AssetDownloadException(message)
    class IntegrityMismatch(val fileName: String) : AssetDownloadException("SHA-256 mismatch for $fileName")
    class InvalidManifest(message: String) : AssetDownloadException(message)
}

File: ios/Runner/AssetDownload/AssetDownloadError.swift
import Foundation

/// Generic errors for the reusable native asset-download framework (ADR-008).
/// Intentionally independent of any specific consumer's error codes (e.g. Tajweed's
/// `TajweedErrorCode`) so this framework has zero knowledge of AI inference,
/// CoreML, or Tajweed — consumers translate these into their own error contracts
/// at the adapter boundary (see `TajweedAssetSync`).
enum AssetDownloadError: Error, Equatable {
  case network(String)
  case badStatusCode(Int)
  case decodeFailed(String)
  case timeout
  case cancelled
  case insufficientStorage(requiredBytes: Int64, availableBytes: Int64)
  case diskError(String)
  case integrityMismatch(fileName: String)
  case invalidManifest(String)
}
